[WIP] Lock Support

miniconfig opened the pull request:
This is my initial attempt at support locks in HomeAssistant.  The backend stuff does seem to work (although there are a few things to clean up), but I wanted some input before I tackle the frontend.

I used "locked" and "unlocked" for the states as opposed to "on" and "off", because I was concerned the latter could be confusing in a potentially dangerous way - but looking at the frontend code (with my non-existent java script experience) it looks like it would have automatically configured a toggle-able card if it were "on" and "off". 

_Should_ this be on/off? Or should I modify the toggle criteria or create something more lock-specific?

miniconfig committed:
Added support for Locks, including those connected through a wink hub.

balloob commented:
Let's keep it at locked / unlocked. We should treat the lock as a new device.

balloob commented on homeassistant/components/lock/__init__.py:
```diff
@@ -0,0 +1,124 @@
+        """ Is the lock locked or unlocked. """
+        return None
+
+    @property
```
No need to implement these

balloob commented on homeassistant/components/lock/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    return True
+
+
+class LockDevice(ToggleEntity):
```
Please implement the methods `lock` and `unlock` (remove the `do_` prefix) and have them `raise NotImplementedError()`

balloob replied:
Can you add a state property?
```python
@property
def state(self):
    locked = self.locked
    if locked is None:
        return STATE_UNKNOWN
    return STATE_LOCKED if locked else STATE_UNLOCKED
```

balloob replied:
Can you not extend the `ToggleEntity` but just `Entity` ?

balloob commented on homeassistant/components/lock/__init__.py:
```diff
@@ -0,0 +1,124 @@
+
+        for lock in target_locks:
+            if service.service == SERVICE_LOCK:
+                lock.do_lock()
```
Let's remove the `do_` prefix.

balloob commented on homeassistant/components/lock/__init__.py:
```diff
@@ -0,0 +1,124 @@
+        return None
+
+    @property
+    def state_attributes(self):
```
This property is not needed right now as the locks do not have extra attributes.

miniconfig committed:
Changed do_lock and do_unlock methods to lock and unlock.
Implemented state method.
Fixed locked method for demo interface.
Changed LockDevice to extend Entity instead of ToggleEntity

miniconfig commented:
I *think* I addressed all of the comments and fixed a few additional things.  I haven't changed the wink library dependency yet, and I'm still working on the tests and frontend, but I think the backend should work.

balloob commented on homeassistant/components/lock/__init__.py:
```diff
@@ -0,0 +1,109 @@
+    return True
+
+
+class LockDevice(Entity):
```
Can you add `lock` and `unlock` methods to this class and `raise NotImplementedError` ?

```python
    def lock(self):
        """ Locks the lock. """
        raise NotImplementedError()
```

miniconfig replied:
I swear I did that...I must be losing my mind.  Coming right up.

balloob commented on homeassistant/components/lock/__init__.py:
```diff
@@ -0,0 +1,109 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+def locked(hass, entity_id=None):
```
Can you rename this to `is_locked` to be in line with `is_on` for light.

balloob commented on homeassistant/components/lock/wink.py:
```diff
@@ -0,0 +1,68 @@
+from homeassistant.components.lock import LockDevice
+from homeassistant.const import CONF_ACCESS_TOKEN
+
+REQUIREMENTS = ['https://github.com/balloob/python-wink/archive/'
```
We should point this at the python-wink version with the lock support. We need to update this too for `wink.py`, `light/wink.py`, `switch/wink.py`, `sensor/wink.py`.
```python
REQUIREMENTS = ['https://github.com/balloob/python-wink/archive/'
                '42fdcfa721b1bc583688e3592d8427f4c13ba6d9.zip'
                '#python-wink==0.2']
```
```

miniconfig committed:
Changed locked method of lock support to "is_locked".
Added lock and unlock methods
Updated wink components to use the new version of the wink library.

miniconfig committed:
Updated wink library version in requirements_all.txt

balloob commented:
Merged :+1:
